application/src/main/java/org/opentripplanner/transit/service/TransitService.java

```diff
@@ -64,6 +64,16 @@
  *   copy-on-write and shares a lot of objects with any other TransitLayer instances.
  */
 public interface TransitService {
+  /**
+   * @return empty if the trip doesn't exist in the timetable (e.g. real-time added)
+   */
+  Optional<List<TripTimeOnDate>> getScheduledTripTimes(Trip trip);
```

Comment:
A (scheduled or RT addd) trip always has a pattern and scheduled trip-times. For RT added trips the scheduled trip-times are generated. However, FlexTrips does not have a pattern/scheduled trip-times. So, I suggest we throw an `IllegalArgumentException` here instead of using optional. The return value should never be `null` or empty. 

Can you update the comment as well?

Reply:
The scheduled trip time is really `null` if the trip is RT added. The reason is because the trip is only added into the snapshot, but not in the base static timetable, while a schedule query is always done on the base static timetable.

This distinction is important for the correct handling of replacement trips, where the scheduled trip time of the replacement can be different from the scheduled trip time of the original, which I have added a test specifically to check this.

Reply:
A `FlexTrip` is not a `Trip` in our model so it won't happen.

Reply:
Still, optional of List is a bit of a strange type.